x-pack/plugins/maps/public/components/single_field_select.tsx

```diff
@@ -0,0 +1,118 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import _ from 'lodash';
+import React from 'react';
+
+import {
+  EuiComboBox,
+  EuiComboBoxProps,
+  EuiComboBoxOptionOption,
+  EuiHighlight,
+  EuiFlexGroup,
+  EuiFlexItem,
+  EuiToolTip,
+} from '@elastic/eui';
+import { IFieldType } from 'src/plugins/data/public';
+import { FieldIcon } from '../../../../../src/plugins/kibana_react/public';
+
+function fieldsToOptions(
+  fields?: IFieldType[],
+  isFieldDisabled?: (field: IFieldType) => boolean
+): Array<EuiComboBoxOptionOption<IFieldType>> {
+  if (!fields) {
+    return [];
+  }
+
+  return fields
+    .map((field) => {
+      const option: EuiComboBoxOptionOption<IFieldType> = {
+        value: field,
+        label: field.name,
+      };
+      if (isFieldDisabled && isFieldDisabled(field)) {
+        option.disabled = true;
+      }
+      return option;
+    })
+    .sort((a, b) => {
+      return a.label.toLowerCase().localeCompare(b.label.toLowerCase());
+    });
+}
+
+type Props = Omit<
+  EuiComboBoxProps<IFieldType>,
+  'isDisabled' | 'onChange' | 'options' | 'renderOption' | 'selectedOptions' | 'singleSelection'
+> & {
+  fields?: IFieldType[];
+  onChange: (fieldName?: string) => void;
+  value?: string; // index pattern field name
+  isFieldDisabled?: (field: IFieldType) => boolean;
+  getFieldDisabledReason?: (field: IFieldType) => string | null;
```

Comment:
not sure if it's worth the effort, but all these optional-params seem mainly to avoid not having to update where `SingleFieldSelect` is being used. This component would be a lot cleaner without all the optional interfaces. (see other comment).

Reply:
There are times when isFieldDisabled and getFieldDisabledReason are not used. For example, when we are showing the field selection for joins. I don't think it makes sense to require all consumers to pass do nothing implementations. I think its cleaner to just make them optional.